pydantic_ai_slim/pydantic_ai/output.py

```diff
@@ -0,0 +1,289 @@
+    """Marker class to use a prompt to tell the model what to output and optionally customize the prompt.
+
+    Example:
+    ```python
+    from pydantic import BaseModel
+
+    from pydantic_ai import Agent, PromptedStructuredOutput
+
+
+    class Vehicle(BaseModel):
+        name: str
+        wheels: int
+
+
+    class Device(BaseModel):
+        name: str
+        kind: str
+
+
+    agent = Agent(
+        'openai:gpt-4o',
+        output_type=PromptedStructuredOutput(
+            [Vehicle, Device],
+            name='Vehicle or device',
+            description='Return a vehicle or device.'
+        ),
+    )
+    result = agent.run_sync('What is a MacBook?')
+    print(repr(result.output))
+    #> Device(name='MacBook', kind='laptop')
+
+    agent = Agent(
+        'openai:gpt-4o',
+        output_type=PromptedStructuredOutput(
+            [Vehicle, Device],
+            template='Gimme some JSON: {schema}'
+        ),
+    )
+    result = agent.run_sync('What is a Ford Explorer?')
+    print(repr(result.output))
+    #> Vehicle(name='Ford Explorer', wheels=4)
+    ```
+    """
+
+    outputs: Sequence[OutputTypeOrFunction[OutputDataT]]
+    name: str | None
+    description: str | None
+    template: str | None
+    """Template for the prompt passed to the model.
+    The '{schema}' placeholder will be replaced with the output JSON schema.
+    If not specified, the default template specified on the model's profile will be used.
+    """
+
+    def __init__(
+        self,
+        type_: OutputTypeOrFunction[OutputDataT] | Sequence[OutputTypeOrFunction[OutputDataT]],
+        *,
+        name: str | None = None,
+        description: str | None = None,
+        template: str | None = None,
+    ):
+        self.outputs = _flatten_output_spec(type_)
+        self.name = name
+        self.description = description
+        self.template = template
+
+
+@dataclass
+class TextOutput(Generic[OutputDataT]):
+    """Marker class to use text output for an output function taking a string argument.
+
+    Example:
+    ```python
+    from pydantic_ai import Agent, TextOutput
+
+
+    def split_into_words(text: str) -> list[str]:
+        return text.split()
+
+
+    agent = Agent(
+        'openai:gpt-4o',
+        output_type=TextOutput(split_into_words),
+    )
+    result = agent.run_sync('Who was Albert Einstein?')
+    print(result.output)
+    #> ['Albert', 'Einstein', 'was', 'a', 'German-born', 'theoretical', 'physicist.']
+    ```
+    """
+
+    output_function: TextOutputFunction[OutputDataT]
+
+
+def _get_union_args(tp: Any) -> tuple[Any, ...]:
+    """Extract the arguments of a Union type if `output_type` is a union, otherwise return an empty tuple."""
+    if typing_objects.is_typealiastype(tp):
+        tp = tp.__value__
+
+    origin = get_origin(tp)
+    if is_union_origin(origin):
+        return get_args(tp)
+    else:
+        return ()
+
+
+def _flatten_output_spec(output_spec: T | Sequence[T]) -> list[T]:
```

Comment:
should these go in `_output.py`?

Reply:
It's used here and in `_output.py`, and currently `_output.py` imports `output.py` but not the other way around, so I don't think we could move it without creating a circular import.